vector/src/Data/Vector/Strict.hs

```diff
@@ -0,0 +1,2675 @@
+  filter, ifilter, filterM, uniq,
+  mapMaybe, imapMaybe,
+  mapMaybeM, imapMaybeM,
+  catMaybes,
+  takeWhile, dropWhile,
+
+  -- ** Partitioning
+  partition, unstablePartition, partitionWith, span, break, spanR, breakR, groupBy, group,
+
+  -- ** Searching
+  elem, notElem, find, findIndex, findIndexR, findIndices, elemIndex, elemIndices,
+
+  -- * Folding
+  foldl, foldl1, foldl', foldl1', foldr, foldr1, foldr', foldr1',
+  ifoldl, ifoldl', ifoldr, ifoldr',
+  foldMap, foldMap',
+
+  -- ** Specialised folds
+  all, any, and, or,
+  sum, product,
+  maximum, maximumBy, maximumOn,
+  minimum, minimumBy, minimumOn,
+  minIndex, minIndexBy, maxIndex, maxIndexBy,
+
+  -- ** Monadic folds
+  foldM, ifoldM, foldM', ifoldM',
+  fold1M, fold1M',foldM_, ifoldM_,
+  foldM'_, ifoldM'_, fold1M_, fold1M'_,
+
+  -- ** Monadic sequencing
+  sequence, sequence_,
+
+  -- * Scans
+  prescanl, prescanl',
+  postscanl, postscanl',
+  scanl, scanl', scanl1, scanl1',
+  iscanl, iscanl',
+  prescanr, prescanr',
+  postscanr, postscanr',
+  scanr, scanr', scanr1, scanr1',
+  iscanr, iscanr',
+
+  -- ** Comparisons
+  eqBy, cmpBy,
+
+  -- * Conversions
+
+  -- ** Lists
+  toList, Data.Vector.Strict.fromList, Data.Vector.Strict.fromListN,
+
+  -- ** Arrays
+  toArray, fromArray, lazyFromArray, toArraySlice, unsafeFromArraySlice, unsafeLazyFromArraySlice,
+
+  -- ** Other vector types
+  G.convert,
+
+  -- ** Mutable vectors
+  freeze, thaw, copy, unsafeFreeze, unsafeThaw, unsafeCopy
+) where
+
+import Data.Vector.Strict.Mutable  ( MVector(..) )
+import Data.Primitive.Array
+import qualified Data.Vector.Fusion.Bundle as Bundle
+import qualified Data.Vector.Generic as G
+
+import Control.DeepSeq ( NFData(rnf)
+#if MIN_VERSION_deepseq(1,4,3)
+                       , NFData1(liftRnf)
+#endif
+                       )
+
+import Control.Monad ( MonadPlus(..), liftM, ap )
+#if !MIN_VERSION_base(4,13,0)
+import Control.Monad (fail)
+#endif
+import Control.Monad.ST ( ST, runST )
+import Control.Monad.Primitive
+import qualified Control.Monad.Fail as Fail
+import Control.Monad.Fix ( MonadFix (mfix) )
+import Control.Monad.Zip
+import Data.Function ( fix )
+
+import Prelude
+  ( Eq, Ord, Num, Enum, Monoid, Functor, Monad, Show, Bool, Ordering(..), Int, Maybe, Either
+  , compare, mempty, mappend, mconcat, return, showsPrec, fmap, otherwise, id, flip, const
+  , (>>=), (+), (-), (<), (<=), (>), (>=), (==), (/=), (&&), (.), ($), seq )
+
+import Data.Functor.Classes (Eq1 (..), Ord1 (..), Read1 (..), Show1 (..))
+import Data.Typeable  ( Typeable )
+import Data.Data      ( Data(..) )
+import Text.Read      ( Read(..), readListPrecDefault )
+import Data.Semigroup ( Semigroup(..) )
+
+import qualified Control.Applicative as Applicative
+import qualified Data.Foldable as Foldable
+import qualified Data.Traversable as Traversable
+
+import qualified GHC.Exts as Exts (IsList(..))
+
+
+-- | Boxed vectors, supporting efficient slicing.
+data Vector a = Vector {-# UNPACK #-} !Int
```

Comment:
Could it be a newtype over lazy boxed vectors? I imagine it would save us some instance definitions.

Reply:
That should be possible. Probably most importantly for cheap conversion between strict and lazy vectors. 

But we should be careful with instances. Anything that produces vectors should be implemented in terms of operations over strict vectors.

Reply:
I think it's worth to employ newtypes if possible, to demonstrate clearly that it is the same data structure just with different instances.